torch/csrc/jit/export.cpp

```diff
@@ -756,6 +757,174 @@ std::tuple<std::string, RawDataExportMap> ExportGraph(
+  // string verisions of their file offsets. This also finalizes the file,
+  // and calling serializeTensor after calling this method is illegal.
+  // NOTE: this method mutates the model proto
+  size_t serializeModelProto(::ONNX_NAMESPACE::ModelProto *model_proto) {
+    JIT_ASSERT(!finalized);
+    JIT_ASSERT(cursor % kFieldAlignment == 0);
+    // Keep track of offset into the file for the model proto
+    JIT_ASSERT(key_to_file_offset.count("__MODEL_PROTO") == 0);
+    key_to_file_offset["__MODEL_PROTO"] = cursor;
+
+    // Swap initializer names to string versions of their file offsets
+    auto *graph = model_proto->mutable_graph();
+    for (int i=0; i < graph->initializer_size(); ++i) {
+      auto *init_proto = graph->mutable_initializer(i);
+      if (key_to_file_offset.count(init_proto->doc_string())) {
+        init_proto->set_doc_string(std::to_string(key_to_file_offset[init_proto->doc_string()]));
+      }
+    }
+
+    // Now swap names for tensor valued attributes to string versions of their
+    // file offsets
+    swapTensorAttributeNames(graph);
+
+    auto serialized_proto = model_proto->SerializeAsString();
+
+    // Write serialized proto size
+    size_t serialized_proto_size = serialized_proto.size();
+    write64BitIntegerLittleEndian(serialized_proto_size);
+    padToNextAlignmentBoundary();
+
+    // Write serialized proto
+    std::fwrite(serialized_proto.c_str(), serialized_proto_size, 1, fp);
+    cursor += serialized_proto_size;
+    padToNextAlignmentBoundary();
+
+    finalized = true;
+    return key_to_file_offset.at("__MODEL_PROTO");
+  }
+
+  ~PyTorchFileWriter() {
+    std::fclose(fp);
+  }
+private:
+  FILE *fp;
+  size_t cursor = 0;
+  std::unordered_map<std::string, size_t> key_to_file_offset;
+  bool finalized = false;
+
+  static constexpr uint64_t kFileFormatVersion = 0x1L;
+  static constexpr uint64_t kFileMagicNumber = 0x314843524f545950L; // PYTORCH1
+  static constexpr uint64_t kFieldAlignment = 64L; // 64 byte alignment supports up to AVX512 for mmap
+  static constexpr uint8_t kPadValue = 0xEF;
+
+  void write64BitIntegerLittleEndian(const uint64_t value) {
+    // TODO endian swap on platforms that need it?
+    std::fwrite(&value, 8u, 1u, fp);
+    cursor += 8u;
+  }
+
+  void writePad(const size_t num_bytes) {
+    uint8_t pad_val = kPadValue;
+    // TODO is there a more efficient way to do this?
+    for (size_t i = 0; i < num_bytes; ++i) {
+      std::fwrite(&pad_val, 1u, 1u, fp);
+    }
+    cursor += num_bytes;
+  }
+
+  void padToNextAlignmentBoundary() {
+    size_t next_offset = (cursor + kFieldAlignment) - (cursor % kFieldAlignment);
+    size_t pad_amount = next_offset - cursor;
+    writePad(pad_amount);
+  }
+
+  void writeFileHeader() {
+    write64BitIntegerLittleEndian(kFileMagicNumber);
+    write64BitIntegerLittleEndian(kFileFormatVersion);
+    padToNextAlignmentBoundary();
+  }
+
+  void swapTensorAttributeNames(::ONNX_NAMESPACE::GraphProto *g) {
```

Comment:
This is very fragile and has huge action-at-a-distance problems. Please make an abstract interface that is used to write the tensors instead.